app/Models/Workshop.php

```diff
@@ -70,6 +71,34 @@ public function externs()
         });
     }
 
+    /**
+     * Returns a collection with all functionaries of the workshop in it.
+     */
+    public function functionaries()
+    {
+        return $this->belongsToMany(User::class, 'role_users');
+    }
+
+    /**
+     * Filters functionaries to only include workshop administrators.
+     */
+    public function administrators()
+    {
+        return $this->functionaries->filter(function ($user, $key) {
+            return $user->hasRole(Role::WORKSHOP_ADMINISTRATOR);
+        });
+    }
+
+    /**
+     * Filters functionaries to only include workshop leaders.
+     */
+    public function leaders()
+    {
+        return $this->functionaries->filter(function ($user, $key) {
+            return $user->hasRole(Role::WORKSHOP_LEADER);
+        });
+    }
+
```

Comment:
The administrators and leaders should return relationships, not collection.
You can define the inverted version of roleWorkshops in the User model for each of them separately.